### PR TITLE
ci: Bump rust toolchain to nightly-2024-09-09

### LINKED ISSUE #5045: bump Rust version
we should bump Rust version in the CI. With Rust 1.81 we get `Error` trait in `core` which would be useful in `no_std`

### PR DESCRIPTION
**Pull request by BAStos525**
## Context
Bump rust toolchain to `nightly-2024-09-09` in iroha2 CI.

### Solution
Closes #5045

### Checklist

- [ ] I've read [`CONTRIBUTING.md`](../CONTRIBUTING.md).
- [ ] (optional) I've written unit tests for the code changes.
- [ ] All review comments have been resolved.
- [ ] All CI checks pass.

### CODE REVIEW

**Comment by mversic** [comment]
update `wasm_builder/src/lib.rs` also

**Review by mversic — DISMISSED** [review]

**Comment by DCNick3** [comment]
How about we commit a `rust-toolchain.toml` file so that the developers get the rust version corresponding to what we have in CI right away?

```toml
[toolchain]
channel = "nightly-2024-09-09"
```

File: primitives/src/addr.rs [hunk 1]
```diff
@@ -610,114 +610,114 @@ mod std_compat {
...
+            "192.168.1.0:9019".parse::<SocketAddrV4>().unwrap(),
             SocketAddrV4 {
                 ip: Ipv4Addr([192, 168, 1, 0]),
                 port: 9019
             }
         );
 
         assert_eq!(
-            SocketAddrV4::from_str("192.168.1.1").unwrap_err(),
+            "192.168.1.1".parse::<Ipv4Addr>().unwrap_err(),
```
Comment (DCNick3) [inline primitives/src/addr.rs:694]: You used an incorrect type here, should be `SocketAddrV4`

Comment (mversic) [reply primitives/src/addr.rs]: fixed

**Comment by mversic** [comment]
> How about we commit a `rust-toolchain.toml` file so that the developers get the rust version corresponding to what we have in CI right away?
> 
> ```toml
> [toolchain]
> channel = "nightly-2024-09-09"
> ```

opened #5051

**Review by mversic — DISMISSED** [review]

File: data_model/src/query/predicate/predicate_atoms/domain.rs [hunk 2]
```diff
@@ -21,6 +21,7 @@ use crate::{
 
 /// A predicate that can be applied to a [`Domain`].
 #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+#[allow(unreachable_patterns)]
```
Comment (nxsaken) [inline data_model/src/query/predicate/predicate_atoms/domain.rs:24]:
```suggestion
#[expect(unreachable_patterns)]
```

Replace everywhere

File: data_model/src/query/predicate/predicate_atoms/mod.rs [hunk 3]
```diff
@@ -145,7 +145,10 @@ where
 
 /// A predicate that can be applied to [`Metadata`].
 #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
-pub enum MetadataPredicateBox {}
+pub enum MetadataPredicateBox {
+    // FIX: Remove all `#[allow(unreachable_patterns)]` from all use sites
```
Comment (nxsaken) [inline data_model/src/query/predicate/predicate_atoms/mod.rs:149]: Remove this comment in favor of expects

File: data_model/src/query/predicate/predicate_atoms/domain.rs (hunk 2 above)
Comment (mversic) [reply data_model/src/query/predicate/predicate_atoms/domain.rs]: nice

Comment (mversic) [reply data_model/src/query/predicate/predicate_atoms/domain.rs]: doesn't work though

**Comment by nxsaken** [comment]
The clippy lint expectation is in the wrong place

**Comment by mversic** [comment]
> The clippy lint expectation is in the wrong place

what's the right place? I tried different places

**Comment by nxsaken** [comment]
These work:

```rust
#[expect(unreachable_patterns)]
const fn on_fn() {
    let x = 5;
    match x {
        _y => (),
        5 => (),
    }
}

const fn in_fn() {
    let x = 5;

    #[expect(unreachable_patterns)]
    match x {
        _y => (),
        5 => (), // the unreachable arm fulfills the expectation
    }

    match x {
        _y => (),
        #[expect(unreachable_patterns)]
        5 => (),
    }
}
```

**Comment by mversic** [comment]
> These work, better use the most specific place:
> 
> ```rust
> #[expect(unreachable_patterns)]
> const fn on_fn() {
>     let x = 5;
>     match x {
>         _y => (),
>         5 => (),
>     }
> }
> 
> const fn in_fn() {
>     let x = 5;
> 
>     #[expect(unreachable_patterns)]
>     match x {
>         _y => (),
>         5 => (), // the unreachable arm fulfills the expectation
>     }
> 
>     match x {
>         _y => (),
>         #[expect(unreachable_patterns)]
>         5 => (),
>     }
> }
> ```

sure, but in this case it lint complains on the enum definition, not where it's used

**Review by mversic — APPROVED** [review]

**Review by nxsaken — APPROVED** [review]

**Review by mversic — APPROVED** [review]

**Pull request merged**